cmd/clusterctl/client/cluster/cert_manager.go

```diff
@@ -94,32 +91,34 @@ func (cm *certManagerClient) Images() ([]string, error) {
 	return images, nil
 }
 
-// EnsureWebhook makes sure the cert-manager Web-hook is Available in a cluster:
-// this is a requirement to install a new provider
+// EnsureCertManager makes sure cert-manager is running and its API is available.
+// This is required to install a new provider.
 // Nb. In order to provide a simpler out-of-the box experience, the cert-manager manifest
 // is embedded in the clusterctl binary.
-func (cm *certManagerClient) EnsureWebhook() error {
+func (cm *certManagerClient) EnsureCertManager() error {
 	log := logf.Log
 
-	// Checks if the cert-manager web-hook already exists, if yes, exit immediately
-	hasWebhook, err := cm.hasWebhook()
+	// Gets the cert-manager objects from the embedded assets.
+	objs, err := cm.getManifestObjs()
 	if err != nil {
 		return err
 	}
-	if hasWebhook {
-		return nil
-	}
 
-	// Otherwise install cert-manager
-	log.Info("Installing cert-manager")
-
-	// Gets the cert-manager objects from the embedded assets.
-	objs, err := cm.getManifestObjs()
+	// Skip re-installing cert-manager if all objects already exist
+	allObjectsExist, err := cm.allObjectsExist(ctx, log, objs)
```

Comment:
I'm still wrapping up my mind around this check.
What If I created a cluster with an older version of clusterctl, installing cert-manager v0.11.0, what happens if I run clusterctl init with a new version of clusterctl (running init multiple times is a supported use case, eg. for installing an additional provider).
If I got this right, this will detect some object is missing, and it will install v0.15.2 on top of v0.11.0 without proper cleanup. This makes init get into the business of doing upgrades and I would like to avoid this.

What about instead running only waitForAPIReady to detect cert-manager is there? are there counter-indications?

Reply:
Yeah, I added this check just this morning, but it occurs to me that you're right and it won't work: https://github.com/kubernetes-sigs/cluster-api/pull/3313#discussion_r452082535

+1 to just running `waitForAPIReady`, although we'll need to do it with a 0 timeout or something (i.e. just check once)

Reply:
second tough, this can lead to a deadlock if something goes wrong in the middle of an installation.
allObjectsExist == false, waitForAPIReady == false

Reply:
last bit. eventually, if it required to check if something is already installed, all the cert-manager components are labeled by clusterctl

Reply:
Updated to no longer use `allObjectsExist` - now it just uses `waitForAPIReady`, and if that fails after one attempt, will call `createObj`.

Reply:
resolving this thread as we no longer have this code 😄